### PR TITLE
Extract forecast reference time

### LINKED ISSUE #100: Extract `cubewrite()` forecast reference time functionality
For splitting `cubewrite()` (#12), forecast reference time functionality should be extracted to a separate function:

https://github.com/ACCESS-NRI/um2nc-standalone/blob/92096ce86f3cdd305570e940397b06633ae8777b/umpost/um2netcdf.py#L178-L202

Tasks:
* [ ] Examine test data for idea of real values
* [ ] Extract code block
* [ ] Add unit tests using 
* [ ] Move `except` block to the coord lookups
* [ ] Convert any strings to constants?
  * [ ] `forecast_reference_time`
  * [ ] `proleptic_gregorian`

For now, let's keep `cubewrite()`'s steps the same, merging into `process()` can be a separate refactoring issue.

### PR DESCRIPTION
**Pull request by truth-quark**
First step to doing #100.

This PR is intended for facilitating testing functionality for fixing the forecast reference time.

As of late Sept 2024, work up to commit fd664c7b only provides a skeleton of unit tests. It doesn't meaningfully test any cube time changes. This PR includes some skipped tests as the testing required is unclear.

Work on this PR needs to answer & address the following:
* [ ] Is the `time.units.calendar == 'gregorian'` code branch & testing required?
* [ ] Is the `time.units.calendar == 'proleptic_gregorian'` code branch & testing required?
* [ ] What inputs should be tested?
* [ ] What outputs should be tested?
* [ ] What cube state/attrs should be tested?
* [ ] Is the code branch covered by `test_fix_forecast_reference_time_standard()` required?

These questions will help determine what work is required. Any comments/ideas welcome!

Fixing & testing this function is specified in #141.

### CODE REVIEW

**Comment by truth-quark** [comment]
Adding Martin as a reviewer to help clarify what is required.

File: umpost/um2netcdf.py
Comment (truth-quark) [inline umpost/um2netcdf.py:372]: `1600` should be a configurable parameter?

File: test/test_um2netcdf.py
Comment (CodeGat) [inline test/test_um2netcdf.py:1204]: Whats the significance of this number..?

Comment (truth-quark) [reply test/test_um2netcdf.py]: It looks like it's ~100 years before 1 Jan 1970:
```python
>>> -16382964.0 / 24 / 365  # convert to days & years
-1870.2013698630137
```
I've moved the value into a test fixture where it can be documented.

**Review by CodeGat — APPROVED** [review]
This looks good to me. But note that I'm missing some of the domain knowledge needed to give a more 'conceptual' approval, so probably wait for others to give their review :D

**Comment by blimlim** [comment]
Pulling this section out into a separate function sounds like a good idea!

I'm still trying to understand the original version of the code, but have a couple of comments and questions so far:

It's a bit tricky to follow I think because of the dynamic overwrite of the `PPfile.calendar` method: https://github.com/ACCESS-NRI/um2nc-standalone/blob/d7117296314ce3ab34f0423767dd9e798be92875/umpost/um2netcdf.py#L95-L110

When I load in an ESM1.5 output file independently, e.g. `/g/data/tm70/sw6175/nc_conversion/output_dirs_for_testing/output000/atmosphere/aiihca.paa1jan`, I get `cube.coord("time").units.calendar` is `standard`. However, when I run `um2nc` on it, the ` if time.units.calendar == 'proleptic_gregorian' and refdate.year < 1600:` branch is still executed because of the above overwrite.


>  Is the time.units.calendar == 'gregorian' code branch & testing required?

Because of the `PPfile.calendar` overwrite, I suspect we will never have a cube with a `gregorian` or `standard` calendar, and so I'm wondering if we can remove this branch entirely, or at least skip tests for it?



>  Is the time.units.calendar == 'proleptic_gregorian' code branch & testing required?

I think it would be good to test this branch eventually. Paleoclimate simulations usually use years before `1600`, and so this branch would be used quite regularly.

---

@MartinDix, do you know why the `forecast_reference_time` coordinate is used to determine if the date is before 1600, rather than using the `time` coordinate directly? https://github.com/ACCESS-NRI/um2nc-standalone/blob/d7117296314ce3ab34f0423767dd9e798be92875/umpost/um2netcdf.py#L367-L369

I'm wondering if it's to ensure that the calendar of dump files isn't changed, e.g. https://github.com/ACCESS-NRI/um2nc-standalone/blob/d7117296314ce3ab34f0423767dd9e798be92875/umpost/um2netcdf.py#L388-L390

but thought it would be good to double check!

File: umpost/um2netcdf.py
Comment (blimlim) [inline umpost/um2netcdf.py:81]: We might be able to use the `cf_units` calendar constants that are used by `def pg_calendar` if it's preferable. E.g. `cf_units.CALENDAR_PROLEPTIC_GREGORIAN`, `cf_units.CALENDAR_GREGORIAN`, and `cf_units.CALENDAR_STANDARD`.

File: umpost/um2netcdf.py
Comment (blimlim) [inline umpost/um2netcdf.py:377]: Just noticed this isn't using the added constants (e.g. `PROLEPTIC_GREGORIAN`, or otherwise the `cf_units` alternatives)

**Comment by truth-quark** [comment]
I'll link in some previous work on `convert_propleptic()` in issue #37.
There's [part of a unit test here](https://github.com/ACCESS-NRI/um2nc-standalone/commit/e50a33a6b3f19050643d02fa5d305ee13d160d8b) which might be useful...

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:78]: Indeed, using the `cf_units` constants is preferable to duplicating them. The fixes are in with 7ee59a23.

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:374]: This is updated in 16a3768a.

**Comment by truth-quark** [comment]
Update: following the 9/10/2024 handover meeting, work on this will be split into 2 parts:
1. Extract the time.calendar functions & provide **basic** unit test scaffolding
2. Retrofit testing in another issue & second PR.

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py]: Defer this to #141.

File: umpost/um2netcdf.py
Comment (aidanheerdegen) [inline umpost/um2netcdf.py:388]: I wonder if it wouldn't be better to just check if `cube` has a FORECAST_REFERENCE_TIME coord outside this routine and only call it if so?

Then you could dump the `try`/`except` block.

As it is removing `FORECAST_PERIOD` too then I guess that either needs to be checked for or put in a smaller `try`/`except`

**Comment by truth-quark** [comment]
> First pass, didn't have time to check the tests. Can take another look tomorrow.

This PR has given us several open questions & highlighted that the requirements are not really known. This function partly blocks work to fix structural/architectural repairs too.  To address all these, how is the following?

* Add code comments for all open questions
* Merge the comments & structural only changes via this PR
* Update #100 with details from this PR & leave it open
* Clarify requirements, what UM provides, what ACCESS NRI needs --> add to issue docs
* Use new PR for modifying the timing repair based on #100

How does that sound?

File: umpost/um2netcdf.py
Comment (aidanheerdegen) [inline umpost/um2netcdf.py:410]: I don't think it is necessary to loop through the whole time array. I think it should be ok to just do the `num2date > newdate > date2num` operation on just the first element, work out the difference between this and the original value and apply that offset to the rest of the `tvals` and `tbnds` in two (vectorised) operations.

File: umpost/um2netcdf.py
Comment (aidanheerdegen) [inline umpost/um2netcdf.py:376]: There is a logic error here. It doesn't correctly deal with `cf_units.CALENDAR_GREGORIAN and refdate.year < 1600`. In this case it will silently convert to proleptic gregorian, which is not correct behaviour if the time is actually gregorian.

File: umpost/um2netcdf.py
Comment (aidanheerdegen) [inline umpost/um2netcdf.py:391]: If the calendar is passed as a function argument all the conversion from hours to days could be done in this function, simplifying the logic.

**Review by blimlim — CHANGES_REQUESTED** [review]
I think the modularisation step looks good, and looks like it will behave exactly the same as before.

Setting up the tests and fixtures for this seems very tricky and intricate though, especially due to the `pg_calendar` overwrite. I'm wondering whether it would be easiest/fastest for this PR to just do the extraction into a separate function, and defer all of the tests to #141?

File: umpost/um2netcdf.py
Comment (blimlim) [reply umpost/um2netcdf.py:376]: This one's also affected by the overwrite of the `PPField.calendar()` method, which I think means the only calendars the rest of the code ever encounters is `cf_units.CALENDAR_PROLEPTIC_GREGORIAN`, `cf_units.CALENDAR_365_DAY` or `cf_units.CALENDAR_360_DAY`. 

https://github.com/ACCESS-NRI/um2nc-standalone/blob/2eaa205a48a5b2d1fba9f994dec0099a48d9fe97/umpost/um2netcdf.py#L100-L115

Possibly these are the only three calendars that the UM ever works with – @MartinDix is that correct?

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:1165]: This is a silly test executing the exception handler block. It's missing an assertion to ensure the cube isn't modified, but it might be simpler to delete it. I can't recall why I added it...

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:1169]: This was testing failure of the 2nd coord lookup `time = cube.coord(TIME)`.

The tests in this PR are mostly foundations that still need to be filled out. I'll add comments to flag they need to be expanded.

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:388]: > As it is removing `FORECAST_PERIOD` too then I guess that either needs to be checked for or put in a smaller `try`/`except`

The `iris` function `remove_coord()` does a `coord()` lookup, which raises an exception if the forecast period is missing:
https://github.com/SciTools/iris/blob/cc60a60508b741709ace45b8ea4fc05e5cd5e8a6/lib/iris/cube.py#L1813-L1830

I think we should consider swapping the 2 `remove_coord()` calls. Remove the forecast reference time first, as it's from to exist from the `try` test. then attempt to remove the forecast period. The current code could fail on removing the forecast period, leaving behind the forecast reference time, HOWEVER this might be the desired behaviour.

File: test/test_um2netcdf.py
Comment (blimlim) [reply test/test_um2netcdf.py:1165]: Ah I see. Otherwise I think leaving it in with some extra comments could be ok!

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:122]: This is where the `Mock` class was useful, it could be passed an object & dynamically make an interface from that.

I can prefix helper funcs with `aux_` & add comments to clarify non-iris interface elements. Would that help?

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:1144]: Yep, that correctness fix is in 0925960.

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:1175]: I've added comments in 6e8b5e2f.

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:1179]: Comments have gone in with 6e8b5e2f.

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:419]: The looping code is old style, conflating logic with manual index fiddling etc. I did some work on #37 to use iteration, but didn't fully solve the problem:
https://github.com/ACCESS-NRI/um2nc-standalone/blob/e50a33a6b3f19050643d02fa5d305ee13d160d8b/umpost/um2netcdf.py#L67-L82

File: test/test_um2netcdf.py
Comment (blimlim) [reply test/test_um2netcdf.py:123]: I think that would be helpful!

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:374]: I've flagged this as a TODO in 43e3ba17.

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:133]: The fix is in d3164e3d.

**Comment by truth-quark** [comment]
@blimlim ^ @aidanheerdegen do you want to check over the recent sets of changes & confirm if they cover the discussion threads here? I'll define "cover" as implements smaller fixes OR makes a note for a future fix.

**Comment by blimlim** [comment]
> confirm if they cover the discussion threads here?

Changes look good

Would you be happy to add comments to the `forecast_ref_time_coord` and `time_coord` fixtures to note that their `unit.calendar` attributes will likely have to be changed in the future, since `um2nc` only ever uses the `cf_units.CALENDAR_PROLEPTIC_GREGORIAN`, `cf_units.CALENDAR_360_DAY`, and `cf_units.CALENDAR_365_DAY` calendars?

File: umpost/um2netcdf.py
Comment (aidanheerdegen) [reply umpost/um2netcdf.py]: I've added a link to the above comment in [this issue](https://github.com/ACCESS-NRI/um2nc-standalone/issues/8#issuecomment-2404062840). This should be dealt with separately taking into account other calendar related issues.

**Comment by truth-quark** [comment]
> Would you be happy to add comments to the `forecast_ref_time_coord` and `time_coord` fixtures to note that their `unit.calendar` attributes will likely have to be changed in the future <snip>

That's gone in with 38205281.

**Review by aidanheerdegen — APPROVED** [review]
LFTM

File: test/test_um2netcdf.py
Comment (blimlim) [reply test/test_um2netcdf.py:1158]: Fixme added in [3820528](https://github.com/ACCESS-NRI/um2nc-standalone/pull/118/commits/38205281c3f167b99360fa76bcc6a80905fc8c2d)

**Review by blimlim — APPROVED** [review]
Looks good to me

**Comment by truth-quark** [comment]
> Setting up the tests and fixtures for this seems very tricky and intricate though, especially due to the `pg_calendar` overwrite. I'm wondering whether it would be easiest/fastest for this PR to just do the extraction into a separate function, and defer all of the tests to #141?

Yep, initial attempts at testing were fiddly & nasty to set up the data structures. Approaching in small, logical steps, extracting helper funcs etc should make things better with gradual logic reduction.

**Pull request merged**
